webp/view/src/main/cpp/webp_decode.cpp

```diff
@@ -0,0 +1,199 @@
+
+    // this does not seems to have difference on Nexus 5
+    config.options.use_threads = 1;
+    switch (bufInfo_.format_) {
+        case SurfaceFormat::SURFACE_FORMAT_RGB_565:
+            config.output.colorspace = MODE_RGB_565;
+            break;
+        case SurfaceFormat::SURFACE_FORMAT_RGBA_8888:
+        case SurfaceFormat::SURFACE_FORMAT_RGBX_8888:
+            config.output.colorspace = MODE_RGBA;
+            break;
+        default:
+            assert( 0 );
+            return;
+    }
+    config.output.width = bufInfo_.width_;
+    config.output.height = bufInfo_.height_;
+    config.output.is_external_memory = 1;
+    config.output.private_memory = reinterpret_cast<uint8_t*>(buf_);
+    config.output.u.RGBA.stride = bufInfo_.stride_ * bytePerPix_;
+    config.output.u.RGBA.rgba  = config.output.private_memory;
+    config.output.u.RGBA.size  = config.output.height *
+                                 config.output.u.RGBA.stride;
+
+    status = WebPDecode(buf, len, &config);
+    WebPFreeDecBuffer(&config.output);
+    delete buf;
+
+    // only change state when it is decoded OK. Our small decoder engine will be
+    // in decoding state when error happens
+    assert(status == VP8_STATUS_OK);
+    if (status == VP8_STATUS_OK) {
+        state_ = state_ready;
+    }
+
+    delete worker_;
+    worker_ = nullptr;
+
+    // if we were asked to release while we are busy decoding (at this point,
+    // this is dangling pointer), we perform the release here to complete the
+    // request.
+    if (stopPending_) {
+        stopPending_ = false;
+        delete  this;
+    }
+}
+
+/*
+ * DecodeFrame(void):
+ *     Start to decode a picture into internal frame memory
+ *     The internal memory layout and size are the same as andriod native
+ *     window to save copying when possible.
+ *
+ *     The decoded frames are scaled up/down by webp decoder to fix the display
+ *     window size.
+ */
+bool WebpDecoder::DecodeFrame(void) {
+    if (state_ == state_decoding)
+        return false;
+    pthread_t       worker_;
+    pthread_attr_t  attrib;
+    pthread_attr_init( &attrib);
+    pthread_attr_setdetachstate(&attrib, PTHREAD_CREATE_DETACHED);
+    int status = pthread_create(&worker_, &attrib, ::DecodeFrame, this);
+    assert(status == 0 );
+    pthread_attr_destroy(&attrib);
+
+    state_ = state_decoding;
+    return true;
+}
+
+/*
+ * DestroyDecoder(void):
+ *     Self-delete if decoder is idle
+ *     Otherwise set up a flag to let decoding thread perform self-delete when
+ *     it finishes. Upon returning from the function, the class pointer is invalid
+ *     and should not be used
+ */
+bool WebpDecoder::DestroyDecoder(void) {
+    if (state_ == state_decoding) {
+        stopPending_ = true;
+        return false;
+    }
+
+    delete this;
+    return true;
+}
+
+/*
+ * private destructor prevent object directly call delete
+ */
+WebpDecoder::~WebpDecoder() {
+    if (buf_)  delete [] buf_;
```

Comment:
good point: in real application, those assert() would become "return" and will need a lot of delete.  The downside is that: gnustl_static does not support it, so need to switch to c++_static, which is the risky one for now.